### PR TITLE
Add VPA into Seed

### LINKED ISSUE #715: Add vertical pod autoscaler into seed.
Prometheus has had trouble in the past with scaling. The vertical pod autoscaler can alleviate some of these issues. I propose adding the VPA behind a feature gate. The VPA consists of 3 components:
* vpa-admission-controller
* vpa-recommender
* vpa-updater

These components would run in the garden namespace of the seed. Along with these components a VPA CRD will be created. Then it will be possible to create a VPA object. The VPA has a pod selector and will make recommendations based on that pods usage. 

Any suggestions/ideas for integrating this into the gardener? @rfranzke @adracus @dkistner

### PR DESCRIPTION
**Pull request by wyb1**
**What this PR does / why we need it**:
Adds the vertical pod autoscaler behind a feature gate. The VPA components are added into the seed in the garden namespace. It will now be possible to create VPA CRs. Additionally, a VPA CR is created for each prometheus. 
**Which issue(s) this PR fixes**:
Fixes #715 

**Special notes for your reviewer**:
/cc  @georgekuruvillak @amshuman-kr @swapnilgm 
**Release note**:

```noteworthy operator
If the new `VPA` feature gate is enabled then Gardener will deploy the [Vertical Pod Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler) into the `garden` namespace of every seed cluster.
```
```noteworthy operator
If the new `VPA` feature gate is enabled then the vertical pod autoscaler will provide scaling recommendations for the central Prometheus in the `garden` namespace as well as all shoot-specific Prometheus instances. It will not yet scale them automatically. We might enable this in a future release.
```

### CODE REVIEW

File: charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml]:
```suggestion
apiVersion: {{ include "deploymentversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml:79]: checksum annotation for this secret is missing

File: charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml]:
```suggestion
apiVersion: {{ include "deploymentversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/updater-deployment.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/updater-deployment.yaml]:
```suggestion
apiVersion: {{ include "deploymentversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]:
```suggestion
apiVersion: {{ include "rbacversion" . }}
```

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]: Better name to make it VPA specific?

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml:106]: Add `apps`

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]: Better name to make it VPA specific?

File: charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/vpa-rbac.yaml]: Better name to make it VPA specific?

File: pkg/operation/common/utils.go
Comment (rfranzke) [inline pkg/operation/common/utils.go]:
```suggestion
	if err := k8sClient.Kubernetes().AppsV1().Deployments(namespace).DeleteCollection(
```

File: pkg/operation/common/utils.go
Comment (rfranzke) [inline pkg/operation/common/utils.go]:
```suggestion
				APIVersion: "apps/v1",
```

File: pkg/operation/common/utils.go
Comment (rfranzke) [inline pkg/operation/common/utils.go]: Why is this needed?

File: pkg/operation/common/utils.go
Comment (rfranzke) [inline pkg/operation/common/utils.go]:
```suggestion
	if err := k8sClient.APIExtension().ApiextensionsV1beta1().CustomResourceDefinitions().DeleteCollection(
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/operation/common/utils.go
Comment (wyb1) [reply pkg/operation/common/utils.go]: I'm not sure I understand. Do you mean why do I need delete options or why am I deleting the crds?

Comment (rfranzke) [reply pkg/operation/common/utils.go]: Why do you need the options. You already specify the API version and kind, so why is it needed twice?

Comment (wyb1) [reply pkg/operation/common/utils.go]: I didn't know it worked without the options. I'll change it to pass an empty DeleteOption struct.

Comment (rfranzke) [reply pkg/operation/common/utils.go]: I don't know whether it works without the options. I just wanted you to clarify why it was implemented that way.

**Comment by rfranzke** [comment]
ping

**Review by amshuman-kr — COMMENTED** [review]
Looks good from what I see for ETCD too (apart from suggestions from @rfranzke). We also would be using with `updateMode` `off` to begin with.

But exposing the recommendation as `/metrics` would be of help to observe/evaluate the recommendations. Something along the lines of the [HPA metrics in kube-state-metrics](https://github.com/kubernetes/kube-state-metrics/blob/master/Documentation/horizontalpodautoscaler-metrics.md).

**Comment by wyb1** [comment]
@amshuman-kr To expose the recommendations as metrics we would have to create a custom exporter. I think it would make sense to add in the future.

**Comment by wyb1** [comment]
I rebased and applied code review changes. I still have to add a checksum for the secret.

**Comment by amshuman-kr** [comment]
> To expose the recommendations as metrics we would have to create a custom exporter. I think it would make sense to add in the future.

Yes. Either a custom exporter or a new VPA collector for `kube-state-metrics`. The latter would be preferable in the long run.

I plan to work on the custom exporter after I am done with my current task this week.

**Comment by rfranzke** [comment]
@wyb1 @amshuman-kr etc., has anybody of you taken a deeper look into the implementation of the VPA?
@mvladev pointed out a few issues that he found:
1. The VPA is registering the `MutatingWebhookConfiguration` itself, see [this code](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/admission-controller/config.go#L94-L127). There is no way to disable the self-registration, and to instead install the `MutatingWebhookConfiguration` manually (with maybe a proper pod/namespace label selector). Since the webhook registers on the creating of pods, what happens if we roll a seed and need to recreate the VPA pod itself on another node?
1. The VPA is performing validation as part of the mutating webhook, see [this code](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/admission-controller/logic/server.go#L184-L190). This is a bad behaviour as there might be other webhooks mutating the object, and validation should be performed cleanly via a `ValidatingWebhookConfiguration`.

How do you suggest to proceed? IMHO, especially the first problem mentioned blocks merging this one and getting VPA support anytime soon?!

**Comment by wyb1** [comment]
@rfranzke 
1. There is currently an open and accepted [issue](https://github.com/kubernetes/autoscaler/issues/1561) for this problem. I could see if anyone is actively working on the issue and if not I can look at it. If it is merged then we can continue with VPA integration in the gardener?
2. I can see how this would be a problem but how do we solve this issue? Do we have any other components that use a `ValidatingWebhookConfiguration` so that I can see how it is best solved?

Since the first problem will block merging, I will suggest closing this PR or changing it to WIP until the first problem has been solved. However, since the open issue in the VPA seems to be accepted, hopefully it won't take too long for it to be integrated. Any other suggestions on how to proceed?

**Comment by rfranzke** [comment]
> There is currently an open and accepted [issue](https://github.com/kubernetes/autoscaler/issues/1561) for this problem. I could see if anyone is actively working on the issue and if not I can look at it. If it is merged then we can continue with VPA integration in the gardener?

> Since the first problem will block merging, I will suggest closing this PR or changing it to WIP until the first problem has been solved. However, since the open issue in the VPA seems to be accepted, hopefully it won't take too long for it to be integrated. Any other suggestions on how to proceed?

Yes, IMHO we should be very careful with the current implementation. Maybe someone else has more thoughts on this as well. cc @vlerenc @mvladev @amshuman-kr 

> I can see how this would be a problem but how do we solve this issue? Do we have any other components that use a `ValidatingWebhookConfiguration` so that I can see how it is best solved?

Not sure whether I understand. It's just that you should not do any validation but just mutation with `MutatingWebhookConfiguration`s. Validation is to be performed with `ValidatingWebhookConfiguration`.

**Comment by wyb1** [comment]
> Not sure whether I understand. It's just that you should not do any validation but just mutation with `MutatingWebhookConfiguration`s. Validation is to be performed with `ValidatingWebhookConfiguration`.

I could only find documentation that suggested `ValidatingWebhookConfiguration`s cannot mutate but not that `MutatingWebhookConfiguration`s can never validate. It makes sense to do it the way you mention, but I'm just curious if there is something that I'm missing.

**Comment by amshuman-kr** [comment]
> 1. The VPA is registering the `MutatingWebhookConfiguration` itself, see [this code](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/admission-controller/config.go#L94-L127). There is no way to disable the self-registration, and to instead install the `MutatingWebhookConfiguration` manually (with maybe a proper pod/namespace label selector). Since the webhook registers on the creating of pods, what happens if we roll a seed and need to recreate the VPA pod itself on another node?

I am quite nervous about the `MutatingWebhookConfiguration` approach of VPA in general. I would have preferred that VPA updated the limits on the higher order objects like `replicasets`, `deployments` and `statefulsets` as HPA does rather than work behind those objects using the `MutatingWebhookConfiguration`. 

But there is one thing I like a lot about the VPA over HPA. The `updatePolicy` `off` 😄 

I would recommend that we use that feature to implement our own little controller to update the higher order objects. This way we can avoid registering the `MutatingWebhookConfiguration` of VPA. But I have not tried if the VPA recommendation works smoothly if `vpa-admission-controller` is not deployed at all. I can try it early next week.

**Comment by mvladev** [comment]
> I could only find documentation that suggested `ValidatingWebhookConfiguration`s cannot mutate but not that `MutatingWebhookConfiguration`s can never validate. It makes sense to do it the way you mention, but I'm just curious if there is something that I'm missing.

In K8S admission of a resource has a several steps. When it reaches to the `mutating` part, it fires a **sequential** requests to all webhooks. `Webhook-1` adds `foo`, `Webhook-2` adds `baz` and so.
Then it goes to `validating` part where it fires **parallel** requests to all validating webhooks and if any of those hooks say is not valid then the admission request is rejected.

As you can see if you do a validation in `Webhook-1` which should forbid `baz` it will safely admit the (in practice invalid) resource due to the **sequential** flow.

> I am quite nervous about the `MutatingWebhookConfiguration` approach of VPA in general. I would have preferred that VPA updated the limits on the higher order objects like `replicasets`, `deployments` and `statefulsets` as HPA does rather than work behind those objects using the `MutatingWebhookConfiguration`.
> 
> But there is one thing I like a lot about the VPA over HPA. The `updatePolicy` `off` 😄
>
> I would recommend that we use that feature to implement our own little controller to update the higher order objects. This way we can avoid registering the `MutatingWebhookConfiguration` of VPA. But I have not tried if the VPA recommendation works smoothly if `vpa-admission-controller` is not deployed at all. I can try it early next week.

Isto did that in the beginning, but this behavior was flawed due to several reasons:

- You are limited to only `Deployments` `DaemonSets`  and other resources which you should know about and can actively change the `PodTemplate` in them. What about `MooMoo` resource which creates `moomoo` `Pods`? Does it have a `PodTemplate` at all or it simply chunks out `Pods` with some static resources?
- Some controller might actually reconcile those resources as well (gardener controller-manager for example) and override them frequently

**Comment by amshuman-kr** [comment]
> * You are limited to only `Deployments` `DaemonSets`  and other resources which you should know about and can actively change the `PodTemplate` in them. What about `MooMoo` resource which creates `moomoo` `Pods`? Does it have a `PodTemplate` at all or it simply chunks out `Pods` with some static resources?

I agree. What is missing is a standardisation like the `Scale` subresource for `ResourceRequirements`. But for our purposes (for `prometheus`, `etcd`)  it would be standard objects that contain `PodTemplate`. So, we have a possiblity to side step the custom resource issue. At least for now.

> * Some controller might actually reconcile those resources as well (gardener controller-manager for example) and override them frequently

Again, agreed. But this is a problem with HPA as well already, right? This is also something we are already dealing with for `MachineDeployment`+`ClusterAutoscaler`. So, this is a more general problem than just VPA.

**Comment by amshuman-kr** [comment]
Examples of special handling during gardener reconciliation...
* For [HPA+`kube-apiserver`](https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/controlplane.go#L266-L271).
* For [`MachineDeployment`+`ClusterAutoscaler`](https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/machines.go#L258-L271)

**Comment by vlerenc** [comment]
Well, has someone tried that out already? What happens, if VPA is down? VPA should be generally stable, but it will be down now and then and it would be bad if then the API server shows issues (slow or otherwise).

On the other hand, if it's generally working and we "only" risk slow responses (timeouts) of the seed (not the shoot) API server, maybe that's "acceptable" (timeout, but not failure), because otherwise we have no VPA and that means we have no vertical scaling and don't get better. That's a core feature we need to have: control plane auto-scaling, here Prometheus, but most importantly ETCD.

As for taking the recommendations and doing the updater ourselves, wouldn't it then not make more sense to have a pre-patched/forked VPA (with a patch/fix)? Not sure whether implementing the thing ourselves is yielding stable results any time soon and when we would do that on the higher-level resources, it's even more complex (and has the above discussed drawbacks).

I hope the practical implications of the mutating and validating web hook flow execution are somewhat limited, because also this seems somewhat dirty on VPA side, but then again, we know how we use it (today - with extensibility it may change, but then again, all who integrate will have to face the reality).

In essence, can you please @wyb1 simulate the rolling update behaviour/check what would happen if VPA is gone or how it behaves in these corner cases?

VPA is the only silver lining I see in the field of control plane auto-scaling. We need a solution, because as it stands right now, we have not much else. For our own seeds, but sooner or later also for the shoots (I would bet rather sooner than later). We are all aware of this Sword of Damocles. Not being able to dynamically react to changing workloads and the complaints this will trigger is problematic for Gardener. We have bet everything on Kubernetes, so Kubernetes (hopefully in the form of VPA) must now solve this issue.

**Comment by rfranzke** [comment]
> Well, has someone tried that out already? What happens, if VPA is down? VPA should be generally stable, but it will be down now and then and it would be bad if then the API server shows issues (slow or otherwise).

I checked again what exactly will happen. VPA registers a webhook for `CREATE` operations of `pods`, however, it doesn't specify the `failurePolicy`. So, it will get defaulted to `Ignore`, i.e., if VPA is down then pod creation takes `~30s` longer than usual (because `30s` is the timeout for webhooks, I think). However, the VPA pod itself should come up eventually. If it would have specified `failurePolicy=Fail`, then manual intervention would have been needed. But in this situation it should recover itself, I guess.

> In essence, can you please @wyb1 simulate the rolling update behaviour/check what would happen if VPA is gone or how it behaves in these corner cases?

Yes, please. We should have a good understanding and be sure that everything works fine without requiring ops effort.

> As for taking the recommendations and doing the updater ourselves, wouldn't it then not make more sense to have a pre-patched/forked VPA (with a patch/fix)? Not sure whether implementing the thing ourselves is yielding stable results any time soon and when we would do that on the higher-level resources, it's even more complex (and has the above discussed drawbacks).

I think we could still go ahead and use it as it is today (if @wyb1's tests succeed). @mvladev opinion?
Still, we might want to contribute and influence it, especially wrt to the webhook creation.
I was mainly concerned about the webhook's `failurePolicy`, but as it defaults to `Ignore` it should be OK. The second problem mentioned above (validation performed in mutation phase) is a bad behaviour but it does not block us from using it - still, we might want to contribute here as well.

> VPA is the only silver lining I see in the field of control plane auto-scaling. We need a solution, because as it stands right now, we have not much else. For our own seeds, but sooner or later also for the shoots (I would bet rather sooner than later). We are all aware of this Sword of Damocles. Not being able to dynamically react to changing workloads and the complaints this will trigger is problematic for Gardener. We have bet everything on Kubernetes, so Kubernetes (hopefully in the form of VPA) must now solve this issue.

Sure, totally agreed. I didn't want to vote against VPA in general, I just wanted to mention the potential problems (hopefully none) and raise the awareness here. Generally, we should have a very good understanding of these crucial things that we are deploying into our clusters, so thank's to @mvladev for digging into the code and calling attention to these parts of the code.

**Comment by vlerenc** [comment]
Sure, if we can make VPA better (in the form of contributions), let's do this. After all, we need it to work.

And yes, we must be extra careful with anything we deploy into our clusters, especially if it's an "API server dependency" of sorts.

**Comment by wyb1** [comment]
So far the worst case scenario I've managed to engineer is that the `vpa-admission-controller` cannot start (replicas=0) and while this is happening the kube-apiserver restarts. Once the apiserver is back up pod creation takes `30s` as @rfranzke mentioned above. However, after this first timeout, subsequent pods are created normally without a timeout. 
Another problem is that if the `mutatingwebhookconfiguration` gets deleted there is no way to reconcile it because it is created as part of the `vpa-admission-controller` start up.

**Comment by vlerenc** [comment]
Thanks @wyb1, that sounds rather promising/not as scary as suspected.

Out of curiosity: Why is the timeout only getting us once? What happens if the admission controller is back online, will it be involved in further pod operations? I would have expected this timeout to hit us again and again otherwise what happens if any one fails only once for 30 seconds. Is it then dropped for good/until the API server restarts? That would be bad - in general.

**Comment by wyb1** [comment]
Even if the `vpa-admission-controller` is down and then comes back up it will be able to continue using the webhook.
As to why only the first webhook call actually uses the timout of `30s` and the rest don't? I'm not entirely sure but I suspect that after the first call, the apiserver realizes that there's no endpoint available. Any attempts after that don't need to wait for the timeout because there is no endpoint. The error from the apiserver is:
```
Failed calling webhook, failing open vpa.k8s.io: failed calling admission webhook "vpa.k8s.io": Post https://vpa-webhook.garden.svc:443/?timeout=30s: no endpoints available for service "vpa-webhook"
```
Anyone know if this is the case?

**Comment by vlerenc** [comment]
Ah @wyb1, so it's the readiness probe/endpoint? The moment the readiness probe fails and the endpoint IP(s) are dropped, the API server knows and doesn't blindly try to reach it anymore. It's a shortcut/smart decision based on Kubernetes endpoint logic/behaviour. Nice!

**Comment by wyb1** [comment]
rebased and added checksum annotation for the secret. @rfranzke could you do another review?

File: charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml
Comment (wyb1) [reply charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml:79]: done

File: charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml
Comment (rfranzke) [inline charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml:40]:
```suggestion
{{ toYaml .Values.podAnnotations | indent 6 }}
```

**Review by rfranzke — COMMENTED** [review]
Found just a very small thing, PTAL.
The rest looks good so far. Still need to try it out. Will get back to you.

File: charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml
Comment (wyb1) [reply charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml:40]: checksum isn't added with `indent 6`

Comment (rfranzke) [reply charts/seed-bootstrap/charts/vpa/templates/admission-controller-deployment.yaml:40]: Correct, sorry.

**Comment by rfranzke** [comment]
I tried it out. Just checked out this PR without enabling the feature gate:

Normal shoot reconciliation fails now:
```
task \"Deploying Shoot monitoring stack in Seed\" failed: timed out after 35.471434655s, last error: no matches for kind \"VerticalPodAutoscaler\" in version \"autoscaling.k8s.io/v1beta1\"
```

**Comment by wyb1** [comment]
Sorry about that. The vpa values hierarchy was not correct. Fixed now. 

If VPA is enabled (and was previously disabled/nonexistent) there is an error that occurs 1-3 times until the VPA CR is deployed. 

```
no matches for kind \"VerticalPodAutoscaler\" in version \"autoscaling.k8s.io/v1beta1\"]"
```

How would you recommend going about this? Or is this acceptable?

**Comment by rfranzke** [comment]
Thanks, will check it out.

Regarding the error: Didn't we talk already about this? Wasn't this the one occurring in the seed bootstrap?

**Comment by wyb1** [comment]
Yes that's the one. It's trying to deploy a VPA CR before the VPA CRD is deployed.

**Comment by rfranzke** [comment]
OK, but we concluded already that it's okay, right? It will retry and reconcile anyways.

**Comment by wyb1** [comment]
The VPA has a new version with some changes and I think it makes sense to change the version now before the PR is merged.
With version 0.4.0 the VPA uses a `targetRef` like the HPA instead of using `labelSelector`.

File: charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml
Comment (swapnilgm) [inline charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml]: Any specific reason behind keeping `imagePullPolicy` `always`?

File: charts/seed-bootstrap/charts/vpa/templates/updater-deployment.yaml
Comment (swapnilgm) [inline charts/seed-bootstrap/charts/vpa/templates/updater-deployment.yaml]: Same as above.

File: charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml
Comment (wyb1) [reply charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml]: No specific reason. I should change it to `IfNotPresent`. Thanks!

Comment (wyb1) [reply charts/seed-bootstrap/charts/vpa/templates/recommender-deployment.yaml]: Changed to `IfNotPresent`

File: charts/seed-bootstrap/charts/vpa/templates/updater-deployment.yaml
Comment (wyb1) [reply charts/seed-bootstrap/charts/vpa/templates/updater-deployment.yaml]: Changed to `IfNotPresent`

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Comment by wyb1** [comment]
I rebased, thanks for the review @rfranzke

**Pull request merged**
